dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/HadoopUtils.java

```diff
@@ -446,39 +447,39 @@ public ExecutionStatus getApplicationStatus(String applicationId) throws BaseExc
         if (StringUtils.isEmpty(applicationId)) {
             return null;
         }
+        String result = getResult(applicationId, "app", false);
+        if (result == null) {
+            //may be in job history
+            result = getResult(applicationId, "job", true);
+        }
+        return getExecutionStatus(result);
+    }
 
-        String result;
-        String applicationUrl = getApplicationUrl(applicationId);
-        logger.debug("generate yarn application url, applicationUrl={}", applicationUrl);
-
-        String responseContent = Boolean.TRUE.equals(PropertyUtils.getBoolean(Constants.HADOOP_SECURITY_AUTHENTICATION_STARTUP_STATE, false)) ? KerberosHttpClient.get(applicationUrl) : HttpUtils.get(applicationUrl);
+    private String getResult(String applicationId, String filedName, boolean isHistory) throws BaseException {
+        String url = isHistory ? getJobHistoryUrl(applicationId) : getApplicationUrl(applicationId);
+        logger.debug("generate yarn application url or job history application url, url={}", url);
+        String responseContent = Boolean.TRUE.equals(PropertyUtils.getBoolean(Constants.HADOOP_SECURITY_AUTHENTICATION_STARTUP_STATE, false)) ? KerberosHttpClient.get(url) : HttpUtils.get(url);
         if (responseContent != null) {
             ObjectNode jsonObject = JSONUtils.parseObject(responseContent);
-            if (!jsonObject.has("app")) {
-                return ExecutionStatus.FAILURE;
+            if (!jsonObject.has(filedName)) {
+                return Constants.FAILED;
             }
-            result = jsonObject.path("app").path("finalStatus").asText();
```

Comment:
I think we could add the `UNDEFINED` logic here instead of making such a big change. WDYT?